File: 838-push-dominoes/838-push-dominoes.cpp
class Solution {
public:
  string pushDominoes(string dominoes) {
    queue<int> q;
    for(int i=0;i<dominoes.size();i++)
      if(dominoes[i]=='L' || dominoes[i]=='R')
        q.push(i);
    string tmp = dominoes;
    while(!q.empty()){
      int s = q.size();
      for(int i=0;i<s;i++){
        auto p = q.front();
        q.pop();
        if(p-1>=0 && dominoes[p]=='L' && dominoes[p-1]=='.'){
          if(p-2>=0 && dominoes[p-2]!='R'){
            tmp[p-1] = 'L';
            q.push(p-1);    
          }
          if(p-2<0)
            tmp[p-1] = 'L';
        }
        if(p+1<dominoes.size() && dominoes[p]=='R' && dominoes[p+1]=='.'){
          if(p+2<dominoes.size() && (dominoes[p+2]!='L')) {
            tmp[p+1] = 'R';
            q.push(p+1);
          }         
          if(p+2>=dominoes.size()){
            tmp[p+1] = 'R';
          }
        }
      }
      dominoes = tmp;
    }
    return dominoes;
  }
};

File: 838-push-dominoes/838-push-dominoes.py
class Solution:
  def pushDominoes(self, dominoes: str) -> str:
    q = []
    for i, ch in enumerate(dominoes):
      if(ch == 'L' or ch == 'R'):
        q.append(i)
    while(len(q)):
      tmp = list(dominoes)
      sz = len(q)
      for _ in range(sz):
        top = q.pop(0)
        if(top-1>=0 and dominoes[top-1] == '.' and dominoes[top] == 'L'):
          if((top-2 >= 0 and dominoes[top-2] != 'R') or top-1 == 0):
            tmp[top-1] = 'L'
            q.append(top-1)
        elif(top+1<len(dominoes) and dominoes[top+1] == '.' and dominoes[top] == 'R'):
          if((top+1 == len(dominoes)-1) or (top+2 < len(dominoes) and dominoes[top+2] != 'L')):
            tmp[top+1] = 'R'
            q.append(top+1);
      dominoes = ''.join(tmp)
    return dominoes;